README.md
# Steam Web API [![Build Status](https://travis-ci.org/shawnsilva/steamwebapi.svg?branch=master)](https://travis-ci.org/shawnsilva/steamwebapi)
## README

### Description
This script will connect to the Steam API to retrieve information about
users/games/groups. 

### How to use

#### Install

To install, git clone the repository or download the archive from GitHub.
Then, run `python setup.py install` to have the package installed. After 
finishing the planned features the package will be available via pip.

#### Use

The following example will assume that JSON data is being returned, and you
want access to the data as returned by Valve.

```python
from steamwebapi.api import ISteamUser, IPlayerService, ISteamUserStats
steamuserinfo = ISteamUser()
steamid = steamuserinfo.resolve_vanity_url("profileURL")['response']['steamid']
usersummary = steamuserinfo.get_player_summaries(steamid)['response']['players'][0]
```

The Steam Web API has multiple inferfaces (e.g., ISteamUser, IPlayerService)
that provide different functions. After instantiating one of the interfaces
the functions can be called with the appropriate paramenters. Each function 
returns a string of either json, xml, or vdf (valve data format). This can be
set by the `DEFAULTFORMAT` variable or changed in the paramters of the
function: `steamuserinfo.resolve_vanity_url("profileURL", format="xml")`.

Alternatively, there is a helper function available to build a "user profile".

```python
from steamwebapi import profiles
user_profile = profiles.get_user_profile("VanityURLOrSteamID")

print(vars(user_profile))
```

### Steam Web API Documentation
	* https://developer.valvesoftware.com/wiki/Steam_Web_API
	* http://wiki.teamfortress.com/wiki/WebAPI
	* https://partner.steamgames.com/documentation/community_data

###REQUIREMENTS
* Python
	* 3.5.*
	* 3.4.*
    * 3.3.*
    * 3.2.*
    * 2.7.*
    * 2.6.*


CHANGELOG.md
#Change Log

### November 27, 2015 - v0.1.1
* API key must be set via the environment variable `STEAM_API_KEY`
* When json format is requested, it is now automatically parsed

### July 05, 2013 - v0.1.0
* Initial script creation.

steamwebapi/test/__init__.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #


steamwebapi/test/test_utils.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

import unittest
from steamwebapi.utils import gid_32_to_64_bit

KNOWN_32_BIT_GID = 1606484
KNOWN_64_BIT_GID = 103582791431127892

class TestUtils(unittest.TestCase):
	def test_gid_32_to_64_bit(self):
		converted_id = gid_32_to_64_bit(KNOWN_32_BIT_GID)
		self.assertEqual(converted_id, KNOWN_64_BIT_GID)


steamwebapi/test/test_community_xml.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

import unittest
import xml.etree.ElementTree as ET
from steamwebapi.api import SteamCommunityXML

# Current Steam ID for user 'vanityURL'
STEAM_USER_ID = '76561198148848265'
# Current Steam Group ID for 'Slackware'
STEAM_GROUP_ID = '103582791431127892'

class TestCommunityXml(unittest.TestCase):
	def test_user_xml(self):
		steamcomm = SteamCommunityXML()
		data = steamcomm.get_user_info('vanityURL')
		xml = ET.ElementTree(ET.fromstring(data))
		self.assertEqual(xml.getroot().tag, 'profile')
		self.assertEqual(xml.findtext("steamID64"), STEAM_USER_ID)

	def test_group_xml(self):
		steamcomm = SteamCommunityXML()
		data = steamcomm.get_group_info('slackware')
		xml = ET.ElementTree(ET.fromstring(data))
		self.assertEqual(xml.getroot().tag, 'memberList')
		self.assertEqual(xml.findtext("groupID64"), STEAM_GROUP_ID)


steamwebapi/user.py
#!/usr/bin/env python
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# user.py
# Version: 0.1.0
# By: Shawn Silva (ssilva at jatgam dot com)
# 
# Created: 07/05/2013
# Modified: 07/05/2013
# 
# Copyright (C) 2013  Shawn Silva
# -------------------------------
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
import re
import json
from api import ISteamUser, IPlayerService, ISteamUserStats

class UserInfo:
    VisibilityState = {1 : "Private", 2 : "Friends Only", 3 : "Friends of Friends", 4 : "Users Only", 5 : "Public"}
    PersonaState = {0 : "Offline", 1 : "Online", 2 : "Busy", 3 : "Away", 4 : "Snooze", 5 : "Looking to Trade", 6 : "Looking to Play"}

    def __init__(self):
        self.steamid = None #The user's 64 bit ID
        self._communityvisibilitystate = None #An integer that describes the access setting of the profile
        self.profilestate = None #If set to 1 the user has configured the profile.
        self.personaname = None #User's display name.
        self.lastlogoff = None #A unix timestamp of when the user was last online.
        self.profileurl = None #The URL to the user's Steam Community profile.
        self.avatar = None #A 32x32 image
        self.avatarmedium = None #A 64x64 image
        self.avatarfull = None #A 184x184 image
        self._personastate = None #The user's status
        #The Following may not be present
        self.commentpermission = None #If present the profile allows public comments.
        self.realname = None #The user's real name.
        self.primaryclanid = None #The 64 bit ID of the user's primary group.
        self.timecreated = None #A unix timestamp of the date the profile was created.
        self.loccountrycode = None #ISO 3166 code of where the user is located.
        self.locstatecode = None #Variable length code representing the state the user is located in.
        self.loccityid = None #An integer ID internal to Steam representing the user's city.
        self.gameid = None #If the user is in game this will be set to it's app ID as a string.
        self.gameextrainfo = None #The title of the game.
        self.gameserverip = None #The server URL given as an IP address and port number separated by a colon, this will not be present or set to "0.0.0.0:0" if none is available.

    @property
    def communityvisibilitystate(self):
        """Return the Visibility State of the Users Profile"""
        if self._communityvisibilitystate == None:
            return None
        elif self._communityvisibilitystate in self.VisibilityState:
            return self.VisibilityState[self._communityvisibilitystate]
        else:
            #Invalid State
            return None

    @communityvisibilitystate.setter
    def communityvisibilitystate(self, value):
        self._communityvisibilitystate = value

    @property
    def personastate(self):
        """Return the Persona State of the Users Profile"""
        if self._personastate == None:
            return None
        elif self._personastate in self.PersonaState:
            return self.PersonaState[self._personastate]
        else:
            #Invalid State
            return None

    @personastate.setter
    def personastate(self, value):
        self._personastate = value

def get_user_info(user):
    userinfo = UserInfo()
    steamuser = ISteamUser()
    regex = re.compile('^\d{17}$')
    if regex.match(user):
        userinfo.steamid = user
    else:
        userinfo.steamid = json.loads(steamuser.resolve_vanity_url(user))['response']['steamid']
    usersummary = json.loads(steamuser.get_player_summaries(userinfo.steamid))['response']['players'][0]
    for key in list(usersummary.keys()):
        if isinstance(usersummary[key], int):
            exec('userinfo.' + key + ' = ' + str(usersummary[key]))
        else:
            exec('userinfo.' + key + ' = ' + '"' + usersummary[key] + '"')

    return userinfo

def main():
    user = get_user_info("vanityURLorSteamID")
    attrs = vars(user)
    print(attrs)
    #print ', '.join("%s: %s" % item for item in attrs.items())
if __name__ == "__main__":
    main()

steamwebapi/utils.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

"""Various utilities for steamwebapi and helper classes"""

from collections import namedtuple

class MyEnum(object):
	def __init__(self):
		raise TypeError("Don't instantiate enums.")

class SteamUniverse(MyEnum):
	INVALID = 0
	PUBLIC = 1
	BETA = 2
	INTERNAL = 3
	DEV = 4

class SteamAccountType(MyEnum):
	__account_values = namedtuple('__account_values', ['id','character','url','instance'])
	INVALID = __account_values(**{'id':0,'character':'I','url':None,'instance':None})
	INDIVIDUAL = __account_values(**{'id':0,'character':'U','url':'id','instance':1})
	MULTISEAT = __account_values(**{'id':2,'character':'M','url':None,'instance':None})
	GAMESERVER = __account_values(**{'id':3,'character':'G','url':None,'instance':1})
	ANONGAMESERVER = __account_values(**{'id':4,'character':'A','url':None,'instance':None})
	PENDING = __account_values(**{'id':5,'character':'P','url':None,'instance':None})
	CONTENTSERVER = __account_values(**{'id':6,'character':'C','url':None,'instance':None})
	CLAN = __account_values(**{'id':7,'character':'g','url':'gid','instance':0})
	CHAT = __account_values(**{'id':8,'character':'T','url':None,'instance':None})
	CHAT_CLAN = __account_values(**{'id':8,'character':'c','url':None,'instance':0x00080000})
	CHAT_LOBBY = __account_values(**{'id':8,'character':'L','url':None,'instance':0x00040000})
	P2PSUPERSEEDER = __account_values(**{'id':9,'character':None,'url':None,'instance':None})
	ANONUSER = __account_values(**{'id':10,'character':'a','url':None,'instance':None})

# https://developer.valvesoftware.com/wiki/SteamID
def id_32_to_64_bit(universe, account_type_id, instance, account_id):
	id_64bit = ((universe << 56) | (account_type_id << 52) | (instance << 32) | account_id )
	return id_64bit

def gid_32_to_64_bit(gid):
	id_64bit = id_32_to_64_bit(SteamUniverse.PUBLIC, SteamAccountType.CLAN.id, SteamAccountType.CLAN.instance, gid)
	return id_64bit

steamwebapi/__init__.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

__version__ = "0.1.1"


steamwebapi/api.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Copyright (C) 2013-2015  Shawn Silva
# ------------------------------------
# This file is part of steamwebapi.
#
# steamwebapi is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

"""Steam Web API interaction"""

import sys

PYMAJORVER, PYMINORVER, PYMICROVER, PYRELEASELEVEL, PYSERIAL = sys.version_info
if PYMAJORVER == 3 and PYMINORVER >= 2:
    #Python >= 3.2
    from urllib.request import urlopen
    from urllib.parse import urlencode
elif PYMAJORVER == 2 and (PYMINORVER == 7 or PYMINORVER == 6):
    #Python 2.6 or 2.7
    from urllib2 import urlopen
    from urllib import urlencode
else:
    raise PythonVersionError("Python Version 2.6 or greater required.")

import json
import os
import re

APIKEY = os.environ.get('STEAM_API_KEY')
DEFAULTFORMAT = 'json' #Set to: xml, json, or vdf
DEFAULTLANG = 'en' #Default language

if not APIKEY:
    print("Steam Web API key environment variable not set, and the key wasn't supplied elsewhere.")
    sys.exit(1)

class _SteamWebAPI(object):
    def __init__(self):
        self.apikey = APIKEY
        self.format = DEFAULTFORMAT
        self.language = DEFAULTLANG

    def create_request_url(self, interface, method, version, parameters):
        """Create the URL to submit to the Steam Web API

        interface: Steam Web API interface containing methods.
        method: The method to call.
        version: The version of the method.
        paramters: Parameters to supply to the method.

        """
        if 'format' in parameters:
            parameters['key'] = self.apikey
        else:
            parameters.update({'key' : self.apikey, 'format' : self.format})
        version = "v%04d" % (version)
        url = "http://api.steampowered.com/%s/%s/%s/?%s" % (interface, method,
            version, urlencode(parameters))
        return url

    def retrieve_request(self, url):
        """Open the given url and decode and return the response

        url: The url to open.

        """
        try:
            data = urlopen(url)
        except:
            print("Error Retrieving Data from Steam")
            sys.exit(2)
        return data.read().decode('utf-8')

    def return_data(self, data, format=None):
        """Format and return data appropriate to the requested API format.

        data: The data retured by the api request

        """
        if format is None:
            format = self.format
        if format == "json":
            formatted_data = json.loads(data)
        else: 
            formatted_data = data
        return formatted_data

class ISteamUser(_SteamWebAPI):
    def __init__(self):
        self.interface = 'ISteamUser'
        super(ISteamUser, self).__init__()

    def get_friends_list(self, steamID, relationship='all', format=None):
        """Request the friends list of a given steam ID filtered by role.

        steamID: The user ID
        relationship: Type of friend to request (all, friend)
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'relationship' : relationship}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetFriendsList', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_player_bans(self, steamIDS, format=None):
        """Request the communities a steam id is banned in.

        steamIDS: Comma-delimited list of SteamIDs
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamids' : steamIDS}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetPlayerBans', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_player_summaries(self, steamIDS, format=None):
        """
        Get summaries of steam accounts.

        steamIDS: Comma-delimited list of SteamIDs (max: 100)
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamids' : steamIDS}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetPlayerSummaries', 2,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_user_group_list(self, steamID, format=None):
        """Request a list of groups a user is subscribed to.

        steamID: User ID
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetUserGroupList', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def resolve_vanity_url(self, vanityURL, url_type=1, format=None):
        """Request the steam id associated with a vanity url.

        vanityURL: The users vanity URL
        url_type: The type of vanity URL. 1 (default): Individual profile, 
                    2: Group, 3: Official game group
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'vanityurl' : vanityURL, "url_type" : url_type}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'ResolveVanityUrl', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

class ISteamUserStats(_SteamWebAPI):
    def __init__(self):
        self.interface = 'ISteamUserStats'
        super(ISteamUserStats, self).__init__()

    def get_global_achievement_percentages_for_app(self, gameID, format=None):
        """Request statistics showing global achievements that have been 
        unlocked.

        gameID: The id of the game.
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'gameid' : gameID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface,
            'GetGlobalAchievementPercentagesForApp', 2, parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_global_stats_for_game(self, appID, count, names, startdate,
            enddate, format=None):
        """Request global stats for a give game.

        appID: The app ID
        count: Number of stats to get.
        names: A list of names of stats to get.
        startdate: The start time to gather stats. Unix timestamp
        enddate: The end time to gather stats. Unix timestamp
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {
            'appid' : appID,
            'count' : count,
            'startdate' : startdate,
            'enddate' : enddate
            }
        count = 0
        for name in names:
            param = "name[" + str(count) + "]"
            parameters[param] = name
            count += 1
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetGlobalStatsForGame', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_number_of_current_players(self, appID, format=None):
        """Request the current number of players for a given app.

        appID: The app ID
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'appid' : appID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface,
            'GetNumberOfCurrentPlayers', 1, parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_player_achievements(self, steamID, appID, language=None,
            format=None):
        """Request the achievements for a given app and steam id.

        steamID: Users steam ID
        appID: The app id
        language: The language to return the results in. None uses default.
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'appid' : appID}
        if format is not None:
            parameters['format'] = format
        if language is not None:
            parameters['l'] = language
        else:
            parameters['l'] = self.language
        url = self.create_request_url(self.interface, 'GetPlayerAchievements', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_schema_for_game(self, appID, language=None, format=None):
        """Request the available achievements and stats for a game.

        appID: The app id
        language: The language to return the results in. None uses default.
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'appid' : appID}
        if format is not None:
            parameters['format'] = format
        if language is not None:
            parameters['l'] = language
        else:
            parameters['l'] = self.language
        url = self.create_request_url(self.interface, 'GetSchemaForGame', 2,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_user_stats_for_game(self, steamID, appID, format=None):
        """Request the user stats for a given game.

        steamID: The users ID
        appID: The app id
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'appid' : appID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetUserStatsForGame', 2,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

class IPlayerService(_SteamWebAPI):
    def __init__(self):
        self.interface = 'IPlayerService'
        super(IPlayerService, self).__init__()

    # RecordOfflinePlaytime, requires auth ticket

    def get_recently_played_games(self, steamID, count=0, format=None):
        """Request a list of recently played games by a given steam id.

        steamID: The users ID
        count: Number of games to return. (0 is all recent games.)
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'count' : count}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetRecentlyPlayedGames', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_owned_games(self, steamID, include_appinfo=1,
            include_played_free_games=0, appids_filter=None, format=None):
        """Request a list of games owned by a given steam id.

        steamID: The users id
        include_appinfo: boolean.
        include_played_free_games: boolean.
        appids_filter: a json encoded list of app ids.
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {
                    'steamid' : steamID,
                    'include_appinfo' : include_appinfo,
                    'include_played_free_games' : include_played_free_games
                    }
        if format is not None:
            parameters['format'] = format
        if appids_filter is not None:
            parameters['appids_filter'] = appids_filter
        url = self.create_request_url(self.interface, 'GetOwnedGames', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_steam_level(self, steamID, format=None):
        """Returns the Steam Level of a user.

        steamID: The users ID
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetSteamLevel', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_badges(self, steamID, format=None):
        """Gets badges that are owned by a specific user

        steamID: The users ID
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetBadges', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_community_badge_progress(self, steamID, badgeID, format=None):
        """Gets all the quests needed to get the specified badge, and which are completed.

        steamID: The users ID
        badgeID: The badge we're asking about
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'badgeid' : badgeID}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetCommunityBadgeProgress', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def is_playing_shared_game(self, steamID, appid_playing, format=None):
        """Returns valid lender SteamID if game currently played is borrowed.

        steamID: The users ID
        appid_playing: The game player is currently playing
        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {'steamid' : steamID, 'appid_playing' : appid_playing}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'IsPlayingSharedGame', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

class ISteamWebAPIUtil(_SteamWebAPI):
    def __init__(self):
        self.interface = 'ISteamWebAPIUtil'
        super(ISteamWebAPIUtil, self).__init__()

    def get_server_info(self, format=None):
        """Request the Steam Web API status and time.

        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetServerInfo', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

    def get_supported_API_list(self, format=None):
        """Request a list of APIs that can be accessed with your APIKEY

        format: Return format. None defaults to json. (json, xml, vdf)

        """
        parameters = {}
        if format is not None:
            parameters['format'] = format
        url = self.create_request_url(self.interface, 'GetSupportedAPIList', 1,
            parameters)
        data = self.retrieve_request(url)
        return self.return_data(data, format=format)

class SteamCommunityXML(_SteamWebAPI):
    USER = 0
    GROUP = 1

    def __init__(self):
        super(SteamCommunityXML, self).__init__()

    def create_request_url(self, profile_type, steamID):
        """Create the url to submit to the Steam Community XML feed."""
        regex = re.compile('^\d{17,}$')
        if regex.match(steamID):
            if profile_type == self.USER:
                url = "http://steamcommunity.com/profiles/%s/?xml=1" % (steamID)
            if profile_type == self.GROUP:
                url = "http://steamcommunity.com/gid/%s/memberslistxml/?xml=1" % (steamID)
        else:
            if profile_type == self.USER:
                url = "http://steamcommunity.com/id/%s/?xml=1" % (steamID)
            if profile_type == self.GROUP:
                url = "http://steamcommunity.com/groups/%s/memberslistxml/?xml=1" % (steamID)
        return url

    def retrieve_request(self, url):
        """Open the given url and return the response

        url: The url to open.

        """
        try:
            data = urlopen(url)
        except:
            print("Error Retrieving Data from Steam")
            sys.exit(2)
        return data.read()

    def get_user_info(self, steamID):
        """Request the Steam Community XML feed for a specific user."""
        url = self.create_request_url(self.USER, steamID)
        data = self.retrieve_request(url)
        return self.return_data(data, format='xml')

    def get_group_info(self, steamID):
        """Request the Steam Community XML feed for a specific group."""
        url = self.create_request_url(self.GROUP, steamID)
        data = self.retrieve_request(url)
        return self.return_data(data, format='xml')

def main():
    # Tests
    import json
    steamuser = ISteamUser()
    steamid = steamuser.resolve_vanity_url("vanirtURL")['response']['steamid']
    #jsondata = json.loads(data)
    print(steamid)
    print(steamuser.get_player_summaries(steamid)['response']['players'])

    pserv = IPlayerService()
    print(pserv.get_recently_played_games(steamid))
    print(steamuser.get_user_group_list(steamid))
    steamcomm = SteamCommunityXML()
    print(steamcomm.get_community_info('vanityURL'))
    # sapi = ISteamWebAPIUtil()
    # print(sapi.get_supported_API_list())

if __name__ == "__main__":
    main()
